File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <bits/stdc++.h>
using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
	int ans = 0, idx=0, w=0;
	queue<int> q;

	for(int i=0; i<bridge_length; i++)
		q.push(0);

	while(!q.empty()) {
		ans++;
		w -= q.front();
		q.pop();

		if(idx < truck_weights.size()) {
			if(w + truck_weights[idx] <= weight) {
				w += truck_weights[idx];
				q.push(truck_weights[idx]);
				idx++;
			}
			else q.push(0);
		}
	}

	return ans;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    ans, cw, i = 0, 0, 0
    q = deque(0 for _ in range(bridge_length))

    while q:
        ans += 1
        cw -= q.popleft()

        if i < len(truck_weights):
            if cw + truck_weights[i] <= weight:
                cw += truck_weights[i]
                q.append(truck_weights[i])
                i += 1
            else:
                q.append(0)

    return ans